Fix groups when joining grouped data frames with duplicates (#2330)

davidkretch opened the pull request:
* Fix subset_join to update group column names in attribute
vars when they are duplicate column names and get renamed.

* Add a test to verify that the group column names after
join are all columns in the data frame.

* Fix build_index_cpp to report correct missing group
column name. Currently when a group column name does not
exist in the data frame, it reports a name from the names 
vector (all columns) instead of the vars vector (group columns).

Fixes #2330.

krlmlr commented:
Thanks. Are the group indices updated correctly? We'll need a test for that (unless it exists, I haven't found any), and also a test for the error message buglet you discovered.

@hadley: Is joining a grouped data frame an operation we support?

hadley commented:
I don't know what joining a grouped data frame should do. Probably just preserve the grouping of `x`?

krlmlr commented:
This pull request addresses the case when columns are renamed. I just checked, group indices look ok after an expanding join (replace `df2 <- data.frame(...)` by `df2 <- expand.grid(...)`).

davidkretch commented:
Thanks. I'll add those tests asap.

It looks to me that the existing behavior when joining grouped data frames is to preserve the grouping of `x` -- at the bottom of `subset_join`, it copies `vars` from `x`: 

```
SEXP vars = x.attr("vars");
if (!Rf_isNull(vars))
  out.attr("vars") = vars;
```

davidkretch commented:
I've added tests for:
* expanding join group indices
* missing group column error message
* attribute `vars` in the joined data frame is null when the original df was not grouped, i.e. the new code doesn't introduce any

krlmlr commented on tests/testthat/test-group-indices.R:
```diff
@@ -22,3 +22,10 @@ test_that("group_indices can be used in mutate (#2160)", {
+test_that("group indices are updated correctly for joined grouped data frames (#2330)", {
+  d1 <- data.frame(x = 1:2, y = 1:2) %>% group_by(x, y)
+  d2 <- expand.grid(x = 1:2, y = 1:2)
+  res <- inner_join(d1, d2, by = "x") %>% group_indices()
```
I think testing the group indices as an invariant will make the test clearer (and easily extensible/adaptable).

```r
expect_equal(group_indices(d1), d1$x)

...

res <- inner_join(d1, d2, by = "x")
expect_equal(group_indices(res), res$x)
```

krlmlr commented on tests/testthat/test-joins.r:
```diff
@@ -370,6 +370,20 @@ test_that("joins between factor and character coerces to character with a warnin
+test_that("group column names get updated when they are duplicates (#2330)", {
+  d1 <- data_frame(x = 1:5, y = 1:5) %>% group_by(x, y)
+  d2 <- data_frame(x = 1:5, y = 1:5)
+  res <- inner_join(d1, d2, by = "x")
```
Here, an explicit check for an expected result will give slightly nicer output in case of failure. We can be explicit  about the suffixes the join uses, too.

davidkretch commented:
Hi, thank you for your review. I've updated the pull request. Hopefully any future PRs have fewer issues.

I added logic to give an error if `subset_join` somehow gets an `x` grouped data frame with non-existent group columns.  I'm not sure how to test the error without directly modifying the `vars` attribute though.  Without this check, it would try to access `names[NA_INTEGER]` and hang. I think the alternatives are to give an error, output with no groups, or not check assuming they're always valid.

I also fixed a serious bug in my previous commit, which accidentally modified `x.attr("vars")`  by reference, which is now `const`. Sorry about that.

krlmlr commented:
Thanks for the edits. You can still rip out the column in a grouped data frame:

```r
iris %>% group_by(Species) %>% { .[["Species"]] <- NULL; . }
```

We could forbid this, but creative users will find a way to break it anyway. I think anything that takes (grouped) data frames should check validity of the input, but this is too much for this PR. I'll merge this as is, the safety net you added will protect us from crashes.
